services/s3/src/it/java/software/amazon/awssdk/services/s3/CopySourceIntegrationTest.java

```diff
@@ -0,0 +1,148 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static software.amazon.awssdk.testutils.service.S3BucketUtils.temporaryBucketName;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.internal.handlers.CopySourceInterceptor;
+import software.amazon.awssdk.services.s3.model.BucketVersioningStatus;
+import software.amazon.awssdk.services.s3.model.CopyObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+
+/**
+ * Integration tests for the {@code sourceBucket}, {@code sourceKey}, and {@code sourceVersionId} parameters for
+ * {@link CopyObjectRequest}. Specifically, we ensure that users are able to seamlessly use the same input for both the
+ * {@link PutObjectRequest} key and the {@link CopyObjectRequest} source key (and not be required to manually URL encode the
+ * COPY source key). This also effectively tests for parity with the SDK v1 behavior.
+ *
+ * @see CopySourceInterceptor
+ */
+@RunWith(Parameterized.class)
+public class CopySourceIntegrationTest extends S3IntegrationTestBase {
+
+    private static final String SOURCE_BUCKET_NAME = temporaryBucketName("copy-source-integ-test-src");
+    private static final String DESTINATION_BUCKET_NAME = temporaryBucketName("copy-source-integ-test-dest");
+
+    @Before
+    public void initializeTestData() throws Exception {
+        createBucket(SOURCE_BUCKET_NAME);
+        createBucket(DESTINATION_BUCKET_NAME);
+    }
+
+    @After
+    public void tearDown() {
+        deleteBucketAndAllContents(SOURCE_BUCKET_NAME);
+        deleteBucketAndAllContents(DESTINATION_BUCKET_NAME);
+    }
+
+    @Parameters
+    public static Collection parameters() throws Exception {
+        return Arrays.asList(
+            "simpleKey",
+            "key/with/slashes",
+            "\uD83E\uDEA3",
+            "specialChars/ +!#$&'()*,:;=?@\"",
+            "%20"
+        );
+    }
+
+    private final String key;
+
+    public CopySourceIntegrationTest(String key) {
+        this.key = key;
+    }
+
+    @Test
+    public void copyObject_WithoutExplicitVersion_AcceptsSameKeyAsPut() throws Exception {
```

Comment:
Can any of these integration test cases be converted to build-time tests, now that we know how the service behaves?

Reply:
I did try to capture most of these test cases in `CopySourceInterceptorTest` as well. They have mostly the same input for their `@Parameters`.

Reply:
That's good! Is there a reason to have the integration tests, then? Fewer tests mean faster releases, and less changes for internet hiccups to delay our release.

Reply:
What I really like about the integration tests is that they validate that we can use identical input for both `PutObjectRequest` and `CopyObjectRequest`, as most users would expect (and where users have been burned before). But I understand your point. If you still feel they are redundant I can remove them though.

Reply:
How long do they take to run?

Reply:
On my machine they take about 30 seconds to run. This is primarily because they're creating buckets per parameter. I can speed this up by reusing buckets between parameters.

Reply:
Reusing buckets speeds this up to ~3-4 seconds. I was biasing for simplicity rather than speed before. Will update.

Reply:
Great! Much better. Fewer service calls mean fewer points of failure.